### DISCUSSION TITLE
Binary protocol.

### DISCUSSION THREAD

**Discussion by ArtemIsmagilov** [memcached/memcached#1154] [Q&A] [answered]
Please tell me. Is your binary protocol already outdated and not used? The documentation says that Metacommands completely replace the commands of the binary protocol, and SASL authentication has long been converted to the ASCII protocol. Your Wikipedia confuses me. It says in 2016 that the binary protocol is new and provides new opportunities...

**Comment by dormando** [comment, marked as answer]
If you are going to refer to a website please include the URL. I don't know what "Your wikipedia" means. I do not have any control over the content of wikipedia. The binary protocol was not new in 2016.

The binary protocol was introduced in 2008. It seemed like a good idea but it was not:
- Security issues
- Poor efficiency prevented adoption by large users
- Limited features and poor extensibility
- SASL authentication rarely worked and caused a lot of support headaches.
- Difficult to implement

I introduced the meta protocol in 2019:
- Better efficiency
- More features
- Small command set (get/set/delete/math)
- Extensible
- Simple and easy to implement

SASL is not currently offered by the meta protocol.

**Comment by ArtemIsmagilov** [comment]
https://github.com/memcached/memcached/wiki/Protocols

**Reply by dormando** [reply]
Thanks for pointing this out. I had missed that page: I've updated it.

**Reply by ArtemIsmagilov** [reply]
> Thanks for pointing this out. I had missed that page: I've updated it.


Thanks!

**Comment by ArtemIsmagilov** [comment]
Hello, @dormando , please tell me. Now I'm studying meta commands and understand that they completely replace the get*/gets*/cas/... and other basic commands. Also, as far as I understand, they support UTF8, which already makes these commands quite high-level and do not require taking into account encoding and so on. This leads me to the conclusion - why are you forced to support old commands in the new version when they are outdated in architecture and, moreover, are absolutely replaceable with more advanced implementation methods. This also applies to SASL authentication, which was binary and is already available in ASCII. I assume you are planning to switch to UTF8 encoding as is implemented in other servers, which is quite convenient for developers. It just turns out that in new releases you are using outdated functionality. I would like to point out that I am not at all a professional in these matters, I just want to understand how exactly future releases are planned and why we maintain support for some teams and so on. For example, PyPi has a library that only works on metacommands and it really doesn’t need get/... and so on. Thanks in advance for your attention.

**Reply by dormando** [reply]
I'm having a hard time understanding most of what you're talking about. I don't know what UTF8 has to do with anything.

If you're asking about why we maintain backwards compatibility in the protocol (binary + old ascii) that's a really simple answer: It would be very annoying to users if we didn't do that. Not everyone has the time to deploy newer clients.

Someday we might remove binprot (with ample warning) but that could be 5+ years. Original ascii protocol will likely stay forever.

**Reply by ArtemIsmagilov** [reply]
thank you very much for the answer

**Comment by ArtemIsmagilov** [comment]
https://github.com/memcached/memcached/wiki/MetaCommands
https://pypi.org/project/meta-memcache/
https://github.com/memcached/memcached/blob/b1aefcdf8a265f8a5126e8aa107a50988fa1ec35/doc/protocol.txt#L79
